core/src/providers/google_ai_studio.rs

```diff
@@ -144,50 +146,95 @@ pub struct Content {
+                                function_call: Some(GoogleAIStudioFunctionCall::try_from(fc)?),
+                                function_response: None,
+                            }]
+                        } else {
+                            vec![Part {
+                                text: assistant_msg.content.clone(),
+                                function_call: None,
+                                function_response: None,
+                            }]
+                        }
+                    }
+                };
 
-        let parts = match m.function_calls {
-            Some(ref fcs) => fcs
-                .iter()
-                .map(|fc| {
-                    Ok(Part {
-                        text: m.content.clone(),
-                        function_call: Some(GoogleAIStudioFunctionCall::try_from(fc)?),
-                        function_response: None,
-                    })
+                Ok(Content {
+                    role: String::from("model"),
+                    parts: Some(parts),
                 })
-                .collect::<Result<Vec<Part>>>()?,
-            None => {
-                vec![Part {
-                    text: match m.role {
-                        // System is passed as a Content. We transform it here but it will be removed
-                        // from the list of messages and passed as separate argument to the API.
-                        ChatMessageRole::System => m.content.clone(),
-                        ChatMessageRole::User => m.content.clone(),
-                        ChatMessageRole::Assistant => m.content.clone(),
-                        _ => None,
-                    },
-
+            }
+            ChatMessage::Function(function_msg) => Ok(Content {
+                role: String::from("function"),
+                parts: Some(vec![Part {
+                    text: None,
                     function_call: None,
-                    function_response: match m.role {
-                        ChatMessageRole::Function => {
-                            GoogleAIStudioFunctionResponse::try_from(m).ok()
+                    function_response: GoogleAIStudioFunctionResponse::try_from(function_msg).ok(),
+                }]),
+            }),
+            ChatMessage::User(user_msg) => {
+                let text = match &user_msg.content {
+                    ContentBlock::Mixed(m) => {
+                        let result = m.iter().try_fold(String::new(), |mut acc, content| {
+                            match content {
+                                MixedContent::ImageContent(_) => {
+                                    Err(anyhow!("Vision is not supported for Google AI Studio."))
+                                }
+                                MixedContent::TextContent(tc) => {
+                                    acc.push_str(&tc.text);
+                                    acc.push(' '); // Add space between texts.
```

Comment:
This will add an extraneous space at the end of the string.

I think we should instead filter out image and error if present. Otherwise join the texts with `\n` not ` `

Reply:
Actually on line 210, I trim the final space. Replaced with a new line instead of a space.